src/core/connection.c

```diff
@@ -73,7 +74,8 @@ QuicConnAlloc(
     // partition can be updated accordingly.
     //
     const uint16_t PartitionIndex =
-        IsServer ? Packet->PartitionIndex : QuicLibraryGetCurrentPartition();
+        DesiredPartitionIndex != NULL ? *DesiredPartitionIndex :
+            IsServer ? Packet->PartitionIndex : QuicLibraryGetCurrentPartition();
```

Comment:
Make the server pass in `Packet->PartitionIndex` so you can get rid of the multiple `?` checks.